primitives/arithmetic/src/traits.rs

```diff
@@ -302,3 +307,519 @@ pub trait SaturatedConversion {
+		///     Ok(())
+		/// }
+		///
+		/// fn overflow() -> Result<(), ArithmeticError> {
+		///     FixedI64::from(i64::MIN).ensure_div_int(-1)?;
+		///     Ok(())
+		/// }
+		///
+		/// assert_eq!(extrinsic_zero(), Err(ArithmeticError::DivisionByZero.into()));
+		/// assert_eq!(overflow(), Err(ArithmeticError::Overflow.into()));
+		/// ```
+		fn ensure_div_int<D: FixedPointOperand>(self, d: D) -> Result<D, ArithmeticError> {
+			self.checked_div_int(d).ok_or_else(|| error::division(self, d))
+		}
+	}
+
+	impl<T: FixedPointNumber> EnsureFixedPointNumber for T {}
+
+	/// Similar to [`TryFrom`] but returning an `ArithmeticError` error.
+	pub trait EnsureFrom<T: PartialOrd + Zero + Copy>:
+		TryFrom<T> + PartialOrd + Zero + Copy
+	{
+		/// Performs the conversion returning an `ArithmeticError` if fails.
+		///
+		/// Similar to [`TryFrom::try_from()`] but returning an `ArithmeticError` error
+		///
+		/// ```
+		/// use sp_arithmetic::{traits::EnsureFrom, ArithmeticError};
+		///
+		/// fn overflow() -> Result<(), ArithmeticError> {
+		///     let byte: u8 = u8::ensure_from(256u16)?;
+		///     Ok(())
+		/// }
+		///
+		/// fn underflow() -> Result<(), ArithmeticError> {
+		///     let byte: i8 = i8::ensure_from(-129i16)?;
+		///     Ok(())
+		/// }
+		///
+		/// assert_eq!(overflow(), Err(ArithmeticError::Overflow.into()));
+		/// assert_eq!(underflow(), Err(ArithmeticError::Underflow.into()));
+		/// ```
+		fn ensure_from(other: T) -> Result<Self, ArithmeticError> {
+			Self::try_from(other).map_err(|_| error::equivalent(other))
+		}
+	}
+
+	/// Similar to [`TryInto`] but returning an `ArithmeticError` error.
+	pub trait EnsureInto<T: PartialOrd + Zero + Copy>:
+		TryInto<T> + PartialOrd + Zero + Copy
+	{
+		/// Performs the conversion returning an `ArithmeticError` if fails.
+		///
+		/// Similar to [`TryInto::try_into()`] but returning an `ArithmeticError` error
+		///
+		/// ```
+		/// use sp_arithmetic::{traits::EnsureInto, ArithmeticError};
+		///
+		/// fn overflow() -> Result<(), ArithmeticError> {
+		///     let byte: u8 = 256u16.ensure_into()?;
+		///     Ok(())
+		/// }
+		///
+		/// fn underflow() -> Result<(), ArithmeticError> {
+		///     let byte: i8 = (-129i16).ensure_into()?;
+		///     Ok(())
+		/// }
+		///
+		/// assert_eq!(overflow(), Err(ArithmeticError::Overflow.into()));
+		/// assert_eq!(underflow(), Err(ArithmeticError::Underflow.into()));
+		/// ```
+		fn ensure_into(self) -> Result<T, ArithmeticError> {
+			self.try_into().map_err(|_| error::equivalent(self))
+		}
+	}
+
+	impl<T: TryFrom<S> + PartialOrd + Zero + Copy, S: PartialOrd + Zero + Copy> EnsureFrom<S> for T {}
+	impl<T: TryInto<S> + PartialOrd + Zero + Copy, S: PartialOrd + Zero + Copy> EnsureInto<S> for T {}
+
+	mod error {
+		use super::{ArithmeticError, Zero};
+
+		#[derive(PartialEq)]
+		enum Signum {
+			Negative,
+			Positive,
+		}
+
+		impl<T: PartialOrd + Zero> From<T> for Signum {
+			fn from(value: T) -> Self {
+				if value < Zero::zero() {
```

Comment:
I don't understand well what you want to say here 🥲 . I need to check if this is less than zero.

Reply:
No problem :laughing:  It was related to eventual the removal of the `Copy`, so that this only accepts a `value: &T` instead of `value: T` but probably not worth it.

Reply:
@ggwpez, regarding this topic, is there any reason why the `BaseArithmetic` trait is not `Copy`?

To make the `Ensure` traits automatically work on `BaseArithmetic`, the `BaseArithmetic` should also be `Copy`. The `ensure_ops` could accept a reference instead to avoid this, but I'm curious if maybe adding `Copy` to `BaseArithmetic`, make things easier for the user.

Reply:
>regarding this topic, is there any reason why the BaseArithmetic trait is not Copy?

That would just force every base arithmetic type to be `Copy` without any use (and possibly bloat the WASM).

>To make the Ensure traits automatically work on BaseArithmetic, the BaseArithmetic should also be Copy.

You can just do `Ensure: BaseArithmetic + Copy`. No need to bloat the `BaseArithmetic` trait for that.

Reply:
Sure, `BaseArithmetic + Copy` is the way of doing it. But someone who wants to update from their `check_ops` version to `ensure_ops` would need to pass Copy for all their `BaseArithmetic` bounds. 

My concern is that I think all implementations of `BaseArithmetic` are already `Copy`, I mean, no exists a `BaseArithmetic` that is not `Copy`, and therefore adding `Copy` to that trait comes with no cost.

Reply:
Okay you beat me to it, here it is https://github.com/paritytech/substrate/pull/13043 :stuck_out_tongue: